Patcher edify: Concepts, Architecture, and Installation files (#2315)
gruntwork-io/docs: docs/2.0/docs/patcher/concepts/patches.md

  - name: terrapatch
    version: "0.1.0"

# Steps necessary to resolve breaking changes
steps:
  - name: "<name-of-step>"
    run: <command-to-run>
  - name: "<name-of-second-step>"
    run: <second-command-to-run>
  # etc
```
[View an example patch in the CIS Service Catalog.](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/c3d5ede211fc3230a7d493ceea43622b337ee88a/.patcher/patches/v0.96.4/switch-to-cis-rds-module/patch.yaml)

## Module consumers and module authors

Patcher is designed for two key personas:

> - **Module authors**: Write or update OpenTofu/Terraform modules.

Reviewer: ```suggestion
- **Module authors**: Authors of OpenTofu/Terraform modules.
```

> - **Module consumers**: Use OpenTofu/Terraform modules created by module authors, typically referenced in Terragrunt units (`terragrunt.hcl` files) or directly in Terraform/OpenTofu code.

Reviewer: ```suggestion
- **Module consumers**: Users of OpenTofu/Terraform modules, typically referenced in Terragrunt units (`terragrunt.hcl` files) or directly in Terraform/OpenTofu code.
```

## For module authors

Module authors often need to introduce breaking changes in their modules, which can create a challenging experience for module consumers who must manually update their code. Patcher simplifies this process by allowing module authors to include a patch YAML file that automates these updates.

When a module consumer executes a patch, their code is updated automatically to reflect changes such as adding or renaming variables, updating provider references, or applying other transformations defined in the patch. 

> While module authors have the flexibility to write any command steps for patches—such as validating tool versions or using `sed` for find-and-replace operations—it is **strongly recommended** to use [`terrapatch`](https://github.com/gruntwork-io/terrapatch). Terrapatch is a Gruntwork tool that enables precise updates to Terraform/OpenTofu HCL files.

Reviewer: ```suggestion
While module authors have the flexibility to write any command steps for patches, such as validating tool versions or using `sed` for find-and-replace operations, it is **strongly recommended** to use [`terrapatch`](https://github.com/gruntwork-io/terrapatch). Terrapatch is a Gruntwork tool that enables precise updates to Terraform/OpenTofu HCL files.
```

By investing a small amount of additional time in authoring a patch, module authors enable all module consumers to seamlessly adopt breaking changes without manual intervention.

## For module consumers

> Module consumers typically reference specific versions of OpenTofu/Terraform modules to maintain consistency. Over time, as new versions of modules are released, consumer code can become outdated. In cases where updates introduce breaking changes, the code must be edited to ensure compatibility—this is where Patcher proves valuable.

Reviewer: ```suggestion
Module consumers typically reference specific versions of OpenTofu/Terraform modules to maintain consistency. Over time, as new versions of modules are released, consumer code can become outdated. In cases where updates introduce breaking changes, consuming code must be edited to ensure compatibility—this is where Patcher proves most valuable.
```
